### ISSUE TITLE
Feature: Automatically add detected shows to list

### ISSUE THREAD

**Issue by ConnorKrammer** [erengy/taiga#101]
Here's what I imagine:

1. user adds video files to a folder on Taiga's watchlist
2. Taiga detects added files and does a naive check based on title characteristics
3. if show is recognized, ask user via a prompt whether they would like to add it to one of their lists
4. include a checkbox "do this for future shows" (or maybe just "choose this list for future shows")

Any thoughts?

**Comment by ConnorKrammer** [comment]
Just to add a user story:

I shared this program with a friend because he's into anime. His existing setup was to keep his anime in two file folders: "watched" and "to watch". When he installed Taiga and set both folders as library folders, his first response was "why isn't it working?" He had never bothered with a MAL account because he tracked everything offline, but Taiga made it worthwhile to set up an account. Since his account was brand new, his lists were obviously empty... and so Taiga ignored his files, which was quite confusing to him. He told me that he expected at least a prompt or something to add the existing anime to one of his lists.

To add some weight to his experience, I went through almost the same thing. It's fairly established for media managers (videos, music, books) to detect existing files and track them automatically -- either with a prompt ("Taiga has detected new video files in one of your libraries. Would you like to add it to your list?") or through a configurable default (always goes to "plan to watch"). Hence the confusion.

(MusicBee does this really well -- it adds files from monitored folders to your library automatically, using an optional filter option. Or (and this is what I prefer) it keeps untracked files in a category called your "inbox", from which you can organize things as you see fit.)

It's useful even if I already have an established MAL account. If my friend copies over an anime for me then Taiga handles adding it to my list. Presto, magical UX.

**Comment by rr-** [comment]
Out of curiosity, why do you keep 'watched' folder? Do you keep everything there, or just selected shows you liked?

**Comment by ConnorKrammer** [comment]
Well, in that particular case it was everything I had watched that I wanted to keep. It was only organized that way at the time because I had no better alternative like Taiga, which tracks which shows have been watched and therefore eliminates the need for separation on the filesystem level. Without Taiga it's a pain to keep track of things otherwise -- you're basically duplicating MAL's list feature, but using folders.

That's why the automatic detection is valuable, especially to someone using their own system before migrating to Taiga. Without auto-detect, they have to manually search for every title, select the correct one, and add it to the appropriate list. A detection feature would allow doing this in one fell swoop. It makes the user experience for a first-timer much smoother and more intuitive.

And to clarify, now that I have Taiga I just keep everything in a single library folder and let the program track things for me. It was only a problem for me (and my friend, as it happens -- he was using the same system) beforehand.

**Comment by rr-** [comment]
I imagine it would be cool to have batch add feature, where you paste all the titles into a box, select status (completed/ptw) and it gets added.
Not sure if this should be a request to MAL itself, though.

**Comment by erengy** [comment]
Thanks for the detailed explanation, @ConnorKrammer. Putting automatic detection of new files under monitored directories aside for now, I think it'd be nice for Taiga to provide a "wizard" interface for first-time users, like so:

1. Select service, enter username & password
2. Add library folders
3. Add anime to list

The third step could work in the following fashion:

- Search library folders for filenames with a valid extension
- Discard different episodes of the same anime title
- Discard the ones that are identified as an anime in user's list
- For each anime title, search the active service and store the results
- Display a list of all anime entries, let the user choose
- Add chosen anime entries to user's list

I'm kinda worried that the step we search the active service may take too long and provide unreliable results, though. Taiga is almost definitely going to miss some anime in the process, but still better than not having the feature, I guess.

@rr- I actually keep a `[Completed]` folder as well. I occasionally go through it and delete some series in order to create free space for new ones, so that's the main reason for me.

**Comment by ConnorKrammer** [comment]
How easy is it to grab a list of all anime titles from MAL? I'm thinking you could fetch just the titles (and synonyms, if possible -- though I have a feeling that takes more requests) in plaintext and filter against that. You could then download the rest of the data (number of episodes, pictures) in the background while the user is viewing their results.

Even if that's not possible, your process would still be significantly quicker than the user adding each title manually. It would be very important to display a progress bar (and maybe a text field containing the currently scanned directory) so that they can see what's going on, though. Users are generally more tolerant of such delays when they know it's:
- a) saving them time
- b) a one-shot deal
- c) bonus if they can see it's actually doing work

And if it's a wizard, it would be useful to allow the user to run it on-demand. It also wouldn't be a bad idea to run it once in the background after each startup, to pick up any new titles the user added but forgot to put in their list.

**Comment by adongy** [comment]
I don't think the current MAL api allow users to fetch ALL titles. See http://myanimelist.net/modules.php?go=api . I provides links to search for a title. I guess it's in order to prevent people from scraping the whole site.

**Comment by ConnorKrammer** [comment]
I just skipped through to the end of MAL's list (sorted by score) by changing the `?limit=` parameter in the url. The list ended at 10,647 titles, so I have a feeling that it' not practical. The alternative would be to keep a 24-hour updated list somewhere online by regularly scraping MAL (with a 50 title/page limit it would take about 200 requests) and compare against that.

That's pretty ugly though.

**Comment by erengy** [comment]
Yep, there's no API method for that. As far as I'm aware, the only way to do it in a single request is to create a dummy user account, add all 10,000+ anime to its list, then fetch that user's list as a database dump. Of course, this comes with a couple of disadvantages:

- It's basically a hack. Making a feature rely on a hack is not a good idea.
- It requires someone to keep the list up to date, manually or automatically.
- The same thing has to work for all services, not just MyAnimeList.

In other words, using the search method is still the most reliable option available. It may take a couple of minutes depending on the user's archive, but it's a one-time thing.

**Comment by Setterwing** [comment]
Thinking Taiga could do this was one of the reasons that made me install it as i keep my anime in a external hd and i wasnt looking foward to add 350+ tiitles to my hb account manualy so i found this program called Animu https://forums.hummingbird.me/t/windows-animu-a-sexy-anime-browser/21067 its been descontinuated but still works has i just used it to auto update my library :D here is the github link https://github.com/hypereddie/Animu with a little luck you can sort it into taiga.
